src/core/exec-invoke.c

```diff
@@ -2548,6 +2557,96 @@ static int create_many_symlinks(const char *root, const char *source, char **sym
         return 0;
 }
 
+static int apply_exec_quotas(const char *target_dir, const char *cgroup_path, ExecDirectoryType type, QuotaLimit ql) {
+        assert(target_dir);
+        assert(ql.quota_set);
+
+        _cleanup_close_ int fd = -EBADF;
+        int r = 0;
+
+        fd = xopenat(AT_FDCWD, target_dir, O_RDONLY|O_CLOEXEC|O_NOCTTY|O_NOFOLLOW);
+        if (fd < 0)
+                return log_debug_errno(r, "Exec quotas: Failed to open '%s': %m", target_dir);
+
+        /* Generate candidate project id */
+        static const sd_id128_t k = SD_ID128_ARRAY(e1,4a,79,9b,64,40,41,4a,a8,46,c2,f3,f9,19,4f,01);
+        _cleanup_free_ char *proj_id_plain = NULL;
+        r = asprintf(&proj_id_plain, "%s%d", cgroup_path, type);
+        if (r < 0)
+                return log_debug_errno(r, "Exec quotas: Failed to generate project id for %s: %m", target_dir);
+
+        uint32_t proj_id = PRJID_CLAMP_INTO_QUOTA_RANGE(siphash24(proj_id_plain, strlen(proj_id_plain), k.bytes));
+
+        /* Check if project quotas are supported */
+#define MAX_PROJ_ID_RETRIES 10
+
+        int retries = 0;
+        while (true) {
+                if (retries >= MAX_PROJ_ID_RETRIES)
+                        return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Exec quotas: Failed to generate unique project id for %s: %m", target_dir);
+
+                struct dqblk req;
+                r = quotactl_fd(fd, QCMD_FIXED(Q_GETQUOTA, PRJQUOTA), proj_id, &req);
+                if (r == -ESRCH || ERRNO_IS_NEG_NOT_SUPPORTED(r) || ERRNO_IS_NEG_PRIVILEGE(r)) {
+                        log_warning("Not applying storage quotas: project quotas are not supported for %s", target_dir);
+                        return 0;
+                } else if (r < 0)
+                        return log_debug_errno(r, "Exec quotas: Failed to query disk quota for prjid %u: %m", proj_id);
+
+                if (!FLAGS_SET(req.dqb_valid, QIF_BLIMITS) || req.dqb_bhardlimit <= 0)
+                        break;
+
+                random_bytes(&proj_id, sizeof(proj_id));
+                proj_id = PRJID_CLAMP_INTO_QUOTA_RANGE(proj_id);
```

Comment:
maybe do `proj_id = (uint32_t) (random_u64_range(PROJ_ID_MAX - PROJ_ID_MIN + 1) + PROJ_ID_MIN)` here, becaus it removes the bias